arch/xtensa/core/syscall_helper.c

```diff
@@ -115,7 +115,8 @@ size_t arch_user_string_nlen(const char *s, size_t maxsize, int *err_arg)
 	 * For MPU systems, this would simply results in access errors
 	 * and the exception handler will terminate the thread.
 	 */
-	if (!xtensa_mem_kernel_has_access((void *)s, maxsize, 0)) {
+	if (!xtensa_mem_kernel_has_access(s, maxsize, 0) ||
+	    arch_buffer_validate(s, maxsize, 0)) {
```

Comment:
Why bother checking kernel_has_access() when you're going to do a proper/slow validation vs. the MMU configuration anyway?  What are the circumstances where arch_buffer_validate() returns zero but xtensa_mem_kernel_has_access() returns non-zero?

For MMU systems, in fact (see ptables.c) these are **exactly the same function**, just with a different wrapper for return value convention.